src/qibolab/qubits.py

```diff
@@ -31,9 +32,11 @@ class Qubit:
             send flux pulses to the qubit.
     """
 
+    model_config = ConfigDict(frozen=False)
```

Comment:
I have the impression that `Qubit` can be frozen. If not in this PR, since it still contains the native gates which may need to be updated (eg. by qibocal) then in #983. It only contains the logical channels (not their configs), which I believe never change.

Reply:
In principle, it could: already in #983 it should be created in the platform, and never touched again...

Almost. For as long as the frequencies are in the channels, we need to leave Qibocal (and any user) the freedom to change that even during the execution (i.e. between two of them).

Reply:
> Almost. For as long as the frequencies are in the channels, we need to leave Qibocal (and any user) the freedom to change that even during the execution (i.e. between two of them).

I think the frequencies are in the configs, not the channels. The configs are inheriting the default `Model` from `qibolab.serialize` which I believe is frozen(?). Therefore, the user already cannot change the frequency directly but they can
1. pass a different config when executing something or,
2. change the default config in the `Platform`

which I believe are acceptable solutions.

The `Qubit` seems to be frozen anyway now. It only has a name (frozen) and channels and the channels only have `str` attributes, all of which I believe are frozen. That's why I think we don't need to "unfreeze" it.

While looking on that, I realized that the `Qubit` also has a `mixer_frequencies` property, which I believe is broken as it tries to access attributes that don't exist anymore (not confirmed though, as I have not tried to use it). We should probably drop this.

Reply:
> I think the frequencies are in the configs, not the channels.

Correct

> The configs are inheriting the default `Model` from `qibolab.serialize` which I believe is frozen(?).

Also correct

> which I believe are acceptable solutions.

The problem is what you would if you need *more* frequencies. In that case, you will have to add new channels.

> The `Qubit` seems to be frozen anyway now. It only has a name (frozen) and channels and the channels only have `str` attributes, all of which I believe are frozen. That's why I think we don't need to "unfreeze" it.

All correct. But one of the channels is actually a dictionary (cross resonances), which is a mutable object (opposite to strings).
In principle, this is not preventing `Qubit` from being frozen (which was only needed for natives), but it would be a bit fake: the address of the dictionary won't change, and that's going to make it possible to freeze the object. But the dictionary items would be conceptually part of the `Qubit`, and they can change at will.

We can choose what we prefer. Freeze or unfreeze shouldn't matter that much in this case.

> While looking on that, I realized that the `Qubit` also has a `mixer_frequencies` property, which I believe is broken as it tries to access attributes that don't exist anymore (not confirmed though, as I have not tried to use it). We should probably drop this.

Agreed.

Reply:
> All correct. But one of the channels is actually a dictionary (cross resonances), which is a mutable object (opposite to strings). In principle, this is not preventing `Qubit` from being frozen (which was only needed for natives), but it would be a bit fake: the address of the dictionary won't change, and that's going to make it possible to freeze the object. But the dictionary items would be conceptually part of the `Qubit`, and they can change at will.

You are right, I missed this `dict`. In that case, I am also fine with leaving it as not frozen.

However, I have the impression that allowing users to attach new channels to qubits is not very useful. To execute pulses on these channels they would also have to create the corresponding instrument channel to provide the port information and also register this to the instrument. QM, which is the only working driver, does not provide any interface for this. Of course it's Python so you can still access the `QmController` from the `Platform` and attach the channel there. This would work, however it requires so much knowledge of the internals that I wouldn't call you a "user" anymore.

Not for this PR, but regarding the channel creation business, we should probably decide: we either drop completely and in that case I think we can safely freeze `Qubit`, or we provide a more user-friendly way to do it. Even if we drop, users could still attach new channels in the platform creation if needed, during `Qubit` creation, so that there is no need to mutate them.